### ISSUE TITLE
`FbClient.close()` calls `dlclose()` on libfbclient — process aborts (SIGABRT) ~66 s after `detach()`

### ISSUE THREAD

**Issue by IgorGZd** [hipercompl/fbdb#17]
## Summary

After `FbDb.detach()`, the process reliably aborts with `SIGABRT` about a minute later,
as long as the process keeps running. The delay is remarkably consistent: **66–68 s**, in
every run.

The failure is not in `detach()` itself, but in the `dlclose()` that follows it. Unloading
the Firebird client library runs its global destructors, one of which throws a C++
exception that nobody catches. `dlclose()` itself returns immediately and successfully —
the abort arrives much later, on a thread the library created, which makes the connection
between cause and effect easy to miss.

The crash originates inside libfbclient's own global destructors, so it is not something
the calling code can guard against. Unloading the Firebird client library has a long
history of problems upstream (see below).

## Environment

| | |
|---|---|
| fbdb | **1.6.2** (hosted, sha256 `be02b4ab43bdbc433b3e170335ad70cf40329766e1d758326fbd38eb5bc6964c`) |
| Dart SDK | 3.13.2 (stable), linux_x64 |
| Build mode | reproduced in **both** AOT (`dart compile exe`) and JIT (`dart run`) |
| OS | Ubuntu 24.04.4 LTS, x86_64 |
| **Firebird client** | 5.0.4 — previously also reproduced with 5.0.2 |
| Firebird server | 5.0.4 (LI-V5.0.4.1812), local, over TCP |
| Database | the stock `employee` sample database |

Originally observed on fbdb 1.6.0 / Dart 3.12.1; re-measured on 1.6.2 with identical
results. The relevant line numbers are unchanged between 1.6.0 and 1.6.2.

## What happens

```
detach()
  → fbdbworker.dart:206    handler returns false ("detaching closes the connection")
  → fbdbworker.dart:101    client.close()
  → fbclient_lib.dart:48   lib?.close()     ==  dlclose(libfbclient)
  → returns normally; ~66 s pass
  → __cxa_finalize         Firebird global destructors run on a library thread
  → C++ exception thrown, uncaught
  → abort()  →  SIGABRT (exit code 134)
```

Backtrace (gdb, fbdb 1.6.2, thread `DartWorker`; the main thread is idle in `sleep`):

```
#3  0x00007ffff7c4527e in __GI_raise (sig=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff7c288ff in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff73cf2f6 in ?? () from /lib/x86_64-linux-gnu/libfbclient.so
#6  0x00007ffff73f4701 in ?? () from /lib/x86_64-linux-gnu/libfbclient.so
#7  0x00007ffff73cf15f in ?? () from /lib/x86_64-linux-gnu/libfbclient.so
#8  0x00007ffff7e56b06 in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#9  0x00007ffff7e571f1 in _Unwind_RaiseException () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#10 0x00007ffff73cf42b in ?? () from /lib/x86_64-linux-gnu/libfbclient.so
...
#15 0x00007ffff7319ff7 in ?? () from /lib/x86_64-linux-gnu/libfbclient.so
#16 0x00007ffff7c47382 in __cxa_finalize (d=0x7ffff74dfd20) at ./stdlib/cxa_finalize.c:82
#17 0x00007ffff725ffbc in ?? () from /lib/x86_64-linux-gnu/libfbclient.so
```

An uncaught C++ exception propagating out of a global destructor during `__cxa_finalize`
— nothing above it in the stack belongs to Dart or to FbDb.

## Measurements (fbdb 1.6.2, 2026-09-03)

Exit code 0 = survived the wait, 134 = SIGABRT.

| variant | runs | result |
|---|---|---|
| stock 1.6.2, AOT, `detach()` then wait 100 s | 3 | **0/3 survived** — SIGABRT at 66 s, 67 s, 68 s |
| stock 1.6.2, JIT (`dart run`), same | 2 | **0/2 survived** — SIGABRT at 67 s, 68 s |
| patched (no `dlclose`), AOT, same | 3 | **3/3 survived** |

Earlier, on fbdb 1.6.0 (4 variants × 3 runs, 100 s wait):

| variant | stock 1.6.0 | patched (no `dlclose`) |
|---|---|---|
| attach → query → `detach()` → wait | **0/3 survived** | **3/3 survived** |
| attach → query → `detach()` → `exit(0)` immediately | 3/3 survived | 2/2 survived |
| attach → query → drop reference, force GC → wait | 3/3 survived | — |
| attach → query → keep reference → wait | 3/3 survived | 2/2 survived |

Two things follow:

* `detach()` alone is fine — exiting immediately after it never crashes. The crash needs
  `client.close()` to actually run.
* The `Finalizer` (fbdb.dart:1036) is **not** the trigger: dropping the reference and
  forcing GC, without `detach()`, never crashed.

## Reproduced in plain C, without Dart

The same abort happens in a ~60-line C program that `dlopen()`s `libfbclient`, attaches via
the **legacy ISC API**, detaches, and `dlclose()`s the library — no Dart, no FFI wrapper,
no object-oriented API. Re-measured today against libfbclient 5.0.4:

```
with dlclose      0/3 survived   SIGABRT after 66 s, 66 s, 66 s
without dlclose   3/3 survived
```

Its output shows that `dlclose()` returns straight away:

```
[0s] attach ok
[0s] detach ok
[0s] dlclose...
[0s] dlclose se vratio      <- "dlclose returned"
[0s] cekam 120 s...         <- "waiting 120 s"
                            <- SIGABRT arrives at 66 s
```

So FbDb does nothing wrong beyond calling `dlclose()` — the client library itself does not
tolerate being unloaded. (The C reproducer is `tester/fb_dlclose_repro.c` in our project;
happy to attach it.)

## Why unloading libfbclient is the problem

The Firebird project has known this since 2008:

* [#2096](https://github.com/FirebirdSQL/firebird/issues/2096) — *"atexit() calls in client
  libraries cause segfaults if the libraries are used in dlopen()ed modules"*
* [#2589](https://github.com/FirebirdSQL/firebird/issues/2589) — *"AV in client/embedded
  library when unloading it"* (reported by Alex Peshkoff, Firebird core developer)

Both are closed as fixed in 2.5, but both carry the label `qa: cannot be tested` — there is
no regression test guarding them. A related race is open today:
[#8652](https://github.com/FirebirdSQL/firebird/issues/8652) — *"Prevent race condition when
GlobalObject destruction routine unlocks global mutex"*, which explicitly notes *"v5 is also
affected"* but targets `6.0 Alpha 1`. That matches what we measured: 5.0.2 and 5.0.4 behave
identically, so upgrading the client does not help.

#2589 describes the failure as intermittent ("sometimes does not close successfully"). We
cannot confirm that from our side: **with a fixed setup we saw 100 % reproduction** across
two fbdb versions, two client versions, AOT and JIT, Dart and C. Given the upstream
description, though, testing this with a single run is unwise — and the 66 s delay means a
short wait looks like success.

## Suggested fix

Do not unload the library. In `lib/src/fbclient/fbclient_lib.dart`:

```dart
  /// Closes the dynamic library.
  void close() {
    if (lib != null) {
      // Deliberately NOT calling lib.close(): dlclose() runs __cxa_finalize,
      // and Firebird's global destructors throw there. The Firebird client
      // library is not safe to unload; see FirebirdSQL/firebird#2096 and #2589.
      lib = null;
    }
  }
```

Cost: the client library stays mapped for the lifetime of the process — a few MB, for a
library that is almost always reused anyway. libfbclient also keeps global state and wants
`fb_shutdown()`; even after that, unloading is not safe.

We have been running exactly this patch (via `dependency_overrides` on a local copy of
1.6.0, now 1.6.2) since August, and the crash is gone with no other behaviour change
observed.

## Minimal reproducer

```dart
import 'dart:io';
import 'package:fbdb/fbdb.dart';

Future<void> main() async {
  final db = await FbDb.attach(
    database: '127.0.0.1:employee',
    user: 'SYSDBA', password: 'masterkey',
    options: FbOptions(dbCharset: 'UTF8'),
  );
  final q = db.query();
  await q.openCursor(sql: r'select first 1 * from rdb$database', parameters: []);
  await q.fetchOne();
  await q.close();

  await db.detach();

  // The wait matters: with exit(0) right after detach() the process always
  // survives, because it dies before client.close() ever runs.
  await Future.delayed(const Duration(seconds: 100));
  print('survived');   // never reached: SIGABRT (134) arrives at ~66 s
  exit(0);
}
```

Verified verbatim in a clean scratch project depending on stock `fbdb: 1.6.2` from pub.dev
(no `dependency_overrides`).

## Note on Windows

This is Linux-specific. On Windows `DynamicLibrary.close()` maps to `FreeLibrary()` rather
than `dlclose()`, and we could not reproduce the abort there: the same four variants × 3
runs, with and without the patch, gave **12/12 survived in both columns** (Windows 11,
Firebird client 5.0, 100 s wait). The decisive cell is `detach()` + wait *without* the
patch — 0/3 on Linux, 3/3 on Windows. So Firebird's global destructors do not misbehave on
DLL unload the way they do under `__cxa_finalize`.

We keep the patch on both platforms anyway, since the library should not be unloaded in
either case.

**Comment by ttyrakow** [comment]
Just pushed 1.6.3 to pub.dev with the adjustment you suggested, but with additional guardrail to apply it on non-Windows systems (if libfbclient handles unloading on Windows graciously, let it be unloaded on this platform).

I've considered different options, including a static library handle (loaded once, never unloaded, shared between all `FbClient` instances), but that might go too far - it would be impossible to create different `FbClient` instances with different versions of libfbclient used by each client (it's probably an edge case, but maybe for testing / benchmarking purposes?).

So, I finally decided to go along with your suggestion (haven't come up with anything better :wink:). The need to unload the fbclient library in a real world application seems very unlikely. As you said: if you connect to a DB once, i't likely you're gonna be connecting again.

And the delay / crash was real and painful, so thanks for your detailed report (the thanks I put in the changelog are well deserved).

**Comment by ttyrakow** [comment]
Tested your example with fbdb 1.6.3 on Debian 13 and it "survived" :smile:.

**Comment by IgorGZd** [comment]
Confirmed fixed with fbdb 1.6.3 from pub.dev, no overrides.

Same environment as in the report (Dart 3.13, libfbclient 5.0.4, Firebird 5 on Linux):
attach → query → detach() → wait 100 s, three parallel runs: 3/3 survived.
With 1.6.2 the same runs were 0/3, SIGABRT at ~66 s.

Our production worker is now built against the stock package and the private fork is gone.
Limiting the change to non-Windows platforms makes sense to me: on Windows we measured
FreeLibrary() to be harmless, so nothing is lost there.

Thank you for the quick turnaround, the same-day release and the credit in the changelog.
Feel free to close this one.

**Comment by ttyrakow** [comment]
Closing. 

Additionally, I've checked whether multiple calls to `DynamicLibrary.open()` load the library just once or multiple times. As per [API docs of dart:ffi:DynamicLibrary.open](https://api.dart.dev/dart-ffi/DynamicLibrary/DynamicLibrary.open.html), ffi ensures that a particular dynamic library is loaded only once (even across multiple isolates) in a process, so not unloading the library and creating subsequent instances of `FbClient` should stick to the same libfbclient native library (loaded previously), unless you specify a different libraries in different instantiations of `FbClient`. If so, your process ends up with multiple versions of libfbclient loaded and mapped. But that's definitely an edge case, and multiple libfbclient libs in your process is probably what you wanted anyway.

**Comment by ttyrakow** [comment]
I forgot to mention this (it's off topic, just a clarification). In your code example you attach to the DB like this:

```dart
final db = await FbDb.attach(
    database: '127.0.0.1:employee',
    user: 'SYSDBA', password: 'masterkey',
    options: FbOptions(dbCharset: 'UTF8'),
  );
```

Setting `dbCharset` to `UTF8` when calling `attach` has no effect (as does setting any other charset). Fbdb **always** sets the client-side (i.e. attachment) encoding to UTF-8 (that's virtually the only encoding supported directly by Dart, without depending on third-party packages like iconv - well, there's also Latin-1, so technically UTF-8 is not the only one, but it's the only one that makes sense in a multilingual case). In short: omit `dbCharset` when attaching (specifying it is not an error, but it's dead code and may bring some confusion).

The `dbCharset` option is relevant when you call `createDatabase` and it sets the default encoding used by the **database** being created (the client / attachment encoding will still be UTF-8).

As I said: not relevant in this issue, just a hint. I'll try to make it more clear in the API documentation, to avoid confusing the database charset with the client charset.

**Comment by IgorGZd** [comment]
Thanks for the clarification!

**Comment by ttyrakow** [comment]
Just for your information. 

After analysis I decided that unloading of the native libfbclient will be disabled on all platforms (starting from version 1.7.1). In the 1.7.1 point release some overhaul regarding multiple connections has been made, during which it turned out trying to unload the native library, even if it worked, doesn't make much sense. If you allow multiple simultaneous connections in a single process (and fbdb handles it now reliably (hopefully :wink:)), even trying to unload the library is a bad idea, because if it works, and the OS really unmaps the shared object from the memory, what remains left to all other active connections in the process?

So, all in all, I've fully adopted your suggestion and `lib.close()` **never** gets executed. The worst that might happen is that if you create multiple instances of `FbDb` with different versions of the native libfbclient library specified for different connections, you end up with some memory occupied by the mapped client shared libraries (even if you no longer use some of them). Since this is really an edge case and more hypothetical than practical (why would you want to load multiple different libfbclient libraries in a single process), I don't consider it a serious drawback.

Anyway, I just wanted to thank you again for submitting this issue, because it made me do some deeper analysis of multi-connection scenarios and resulted in significant improvements in this regard, and overall better quality of the fbdb package. Even if multiple connection scenario might not be very common, this is definitely **not** an edge case and should be handled properly. Now it is :smile:

**Comment by IgorGZd** [comment]
Thank you for creating and maintaining FbDb!
